lib/config/param.c

```diff
@@ -907,3 +907,96 @@ param_get(
     return ps->ps_stringify(
         ps, (char *)params->p_params.as_generic + ps->ps_offset, buf, buf_sz, needed_sz);
 }
+
+merr_t
+param_set(
+    const struct params *const     params,
+    const struct param_spec *const pspecs,
+    const size_t                   pspecs_sz,
+    const char *const              param,
+    const char *const              value)
+{
+    merr_t                   err = 0;
+    cJSON *                  node;
+    const struct param_spec *ps = NULL;
+    void *                   old;
+    void *                   data;
+
+    INVARIANT(params);
+    INVARIANT(pspecs);
+    INVARIANT(pspecs_sz > 0);
+    INVARIANT(param);
+    INVARIANT(value);
+
+    node = cJSON_Parse(value);
+    if (!node) {
+        if (cJSON_GetErrorPtr()) {
+            log_err("Failed to parse %s %s: %s", params_logging_context(params), param, value);
+            return merr(EINVAL);
+        } else {
+            return merr(ENOMEM);
+        }
+    }
+
+    for (size_t i = 0; i < pspecs_sz; i++) {
+        if (!strcmp(pspecs[i].ps_name, param)) {
+            ps = &pspecs[i];
+        }
+    }
+
+    if (!ps) {
+        log_err("Unknown parameter %s", param);
+        err = merr(EINVAL);
+        goto out;
+    }
+
+    if (!(ps->ps_flags & PARAM_FLAG_WRITABLE)) {
+        log_err("%s is not wriable", param);
+        err = merr(EINVAL);
+        goto out;
+    }
+
+    assert(ps->ps_size > 0);
+
+    data = (char *)params->p_params.as_generic + ps->ps_offset;
+
+    old = malloc(ps->ps_size);
+    if (!old) {
+        err = merr(errno);
+        goto out;
+    }
+
+    memcpy(old, data, ps->ps_size);
+
+    if (!ps->ps_convert(ps, node, data)) {
+        log_err("Failed to convert %s %s", params_logging_context(params), param);
+        err = merr(EINVAL);
+        goto pop;
+    }
+
+    if (!ps->ps_validate(ps, data)) {
+        log_err("Failed to validate %s %s", params_logging_context(params), param);
+        err = merr(EINVAL);
+        goto pop;
+    }
+
+    if (ps->ps_validate_relations && !ps->ps_validate_relations(ps, params)) {
+        log_err(
+            "Failed to validate parameter relationships for %s %s",
+            params_logging_context(params),
+            ps->ps_name);
+        err = merr(EINVAL);
+        goto pop;
+    }
+
+    goto out;
+
+pop:
+    memcpy(data, old, ps->ps_size);
+    free(old);
```

Comment:
These two memcpy()s are what is creating the transactional approach.

1. Copy the old data from params struct
2. Set the new data
3. Validate it
4. If validation/conversion failed, pop the data back into place

Reply:
Right, but it's not transactional.  There's a window there where someone could see garbage.  All that has to happen is for the validation to fail, the thread loses the CPU, some other reads reads the value.  It could happen even without losing the CPU if two simultaneously running threads collided, albeit that latter is unlikely, but would be impossible to debug..

Reply:
I see what you're saying now. Will fix.

Reply:
Implemented in latest updates, but unless memcpy is atomic, I think there is still a chance to read garbage if another thread is reading the set parameter as data is being copied into it.

I think this is more of a failure from the POV of how components react to changed parameters. Should be some sort of registration system where consumers can be notified of change.

Reply:
You could use a switch on ps_type and do assignments, which are atomic on x86 for scalar types.